File: Dynamic_Programming/1269. Number of Ways to Stay in the Same Place After Some Steps.cpp
/*
1269. Number of Ways to Stay in the Same Place After Some Steps
*/

// TC:O(n^2) SC:O(n)
class Solution {
public:
    int numWays(int steps, int arrLen) {
        vector<long long> dp(min(arrLen, steps),0);
        dp[0] = 1;
        int M = 1e9+7;
        for(int i = 0; i < steps; i++){
            auto dp_tmp = dp;
            for(int j = 0; j < min(arrLen,steps/2+1); j++){
                dp[j] = (dp[j] + (j-1 < 0 ? 0 : dp_tmp[j-1]) + (j+1 >= min(arrLen,steps/2+1) ? 0 :  dp_tmp[j+1])) %M;
            }
        }

        return dp[0];
    }
};

/*
問你走了k步以後你還待在原點的可能性
每一步你可以待在原點也可以向左or向右走一步
n :dp[0] += dp[1]
n-1 : dp[1] = dp_prev[0] + dp_prev[2];
這題的優化就是不關心step/2+1的可能性 因為他肯定走不回來

*/

File: Dynamic_Programming/1269. Number of Ways to Stay in the Same Place After Some Steps.py
# TC:O(n * min(arrLen, n)) SC:O(min(arrLen - 1, steps))
class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        '''
        當前pos = 0, 每次可以停留 or 走左右 問說走了steps次以後可以停留在0的方式有多少種
        代表每個位置他的可能性來自於左+右+當前位置
        '''
        MOD = 10 ** 9 + 7
        mn = min(arrLen - 1, steps)
        f = [0] * (mn + 1)
        f[0] = 1
        for k in range(steps):
            prev, nxt = 0, 0
            farDist = mn if mn < k + 1 else k + 1
            for i in range(0, farDist + 1):
                nxt = f[i+1] if i + 1 <= mn else 0
                prev, f[i] = f[i], (f[i] + prev + nxt) % MOD
        return f[0]
        '''
        遞歸
        @cache
        def dfs(pos:int, k:int) -> int:
            if k == steps:
                return 1 if pos == 0 else 0
            if pos - (steps - k) > 0:
                return 0
            ans1 = dfs(pos + 1, k + 1) if pos + 1 < arrLen else 0
            ans2 = dfs(pos, k + 1)
            ans3 = dfs(pos - 1, k + 1) if pos - 1 >= 0 else 0
            return (ans1 + ans2 + ans3) % MOD
        return dfs(0, 0) % MOD
        '''
